src/System.Data.SqlClient/src/System/Data/Common/DbEnumerator.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Data;
+using System.Data.ProviderBase;
+using System.Diagnostics;
+using System.ComponentModel;
+
+namespace System.Data.Common
+{
+    public class DbEnumerator : IEnumerator
+    {
+        internal DbDataReader _reader;
+        internal DbDataRecord _current;
+        internal SchemaInfo[] _schemaInfo; // shared schema info among all the data records
+        private FieldNameLookup _fieldNameLookup;
+
+        private bool _closeReader;
+
+        public DbEnumerator(DbDataReader reader, bool closeReader)
+        {
+            if (null == reader)
+            {
+                throw ADP.ArgumentNull("reader");
+            }
+            _reader = reader;
+            _closeReader = closeReader;
+        }
+
+
+        public object Current
+        {
+            get
+            {
+                return _current;
+            }
+        }
+
+        /*public IDataRecord Current {
+            get {
+                return _current;
+            }
+        }*/
+
+        /*
+                virtual internal IDataRecord NewRecord(SchemaInfo[] si, object[] values, PropertyDescriptorCollection descriptors) {
+                    return new DbDataRecord(si, values, descriptors);
+                }
+
+                virtual internal void GetValues(object[] values) {
+                    _reader.GetValues(values);
+                }
+        */
+
+        public bool MoveNext()
+        {
+            if (null == _schemaInfo)
+            {
+                BuildSchemaInfo();
+            }
+
+            Debug.Assert(null != _schemaInfo, "unable to build schema information!");
+            _current = null;
+
+            if (_reader.Read())
+            {
+                // setup our current record
+                object[] values = new object[_schemaInfo.Length];
+                _reader.GetValues(values); // this.GetValues()
+                _current = new DataRecordInternal(_schemaInfo, values, _fieldNameLookup); // this.NewRecord()
+                return true;
+            }
+            if (_closeReader)
+            {
+                _reader.Dispose();
```

Comment:
Shouldn't this happen in an IDisposable.Dispose implementation rather than here, in case, for example, the enumerator isn't enumerated to the end?